src/compose/ports.ts

```diff
@@ -0,0 +1,202 @@
+	internalEnd: number;
+	externalStart: number;
+	externalEnd: number;
+	protocol: string;
+	host: string;
+}
+
+export class PortMap {
+
+	private internalStart: number;
+	private internalEnd: number;
+	private externalStart: number;
+	private externalEnd: number;
+	private protocol: string;
+	private host: string;
+
+	public constructor(portStrOrObj: string | PortRange) {
+		if (_.isString(portStrOrObj)) {
+			this.parsePortString(portStrOrObj);
+		} else {
+			this.internalStart = portStrOrObj.internalStart;
+			this.internalEnd = portStrOrObj.internalEnd;
+			this.externalStart = portStrOrObj.externalStart;
+			this.externalEnd = portStrOrObj.externalEnd;
+			this.protocol = portStrOrObj.protocol;
+			this.host = portStrOrObj.host;
+		}
+	}
+
+	public toDockerOpts(): DockerPortOptions {
+		const internalRange = this.generatePortRange(this.internalStart, this.internalEnd);
+		const externalRange = this.generatePortRange(this.externalStart, this.externalEnd);
+
+		const exposed: { [key: string]: {} } = {};
+		const portBindings: PortBindings = {};
+
+		_.zipWith(internalRange, externalRange, (internal, external) => {
+			exposed[`${internal}/${this.protocol}`] = {};
+
+			portBindings[`${internal}/${this.protocol}`] = [
+				{ HostIp: this.host, HostPort: external!.toString() },
+			];
+		});
+
+		return {
+			exposedPorts: exposed,
+			portBindings,
+		};
+	}
+
+	/**
+	 * fromDockerOpts
+	 *
+	 * We need to be able to compare target and running containers, but
+	 * running containers store their ports in a separate format (see the
+	 * output of toDockerOpts).
+	 *
+	 * This function takes the output of toDockerOpts (or more accurately,
+	 * the output of the docker daemon when probed about a given container)
+	 * and produces a list of PortMap objects, which can then be compared.
+	 *
+	 */
+	public static fromDockerOpts(
+		portBindings: PortBindings,
+	): PortMap[] {
+
+		// Create a list of portBindings, rather than the map (which we can't
+		// order)
+		const portMaps = _.map(portBindings, (hostObj, internalStr) => {
+
+			const match = internalStr.match(DOCKER_OPTS_PORTS_REGEX);
+			if (match == null) {
+				throw new Error(`Could not parse docker port output: ${internalStr}`);
+			}
+			const internal = parseInt(match[1], 10);
+			const external = parseInt(hostObj[0].HostPort, 10);
+
+			const host = hostObj[0].HostIp;
+			const proto = match[2] || 'tcp';
+
+			return new PortMap({
+				internalStart: internal,
+				internalEnd: internal,
+				externalStart: external,
+				externalEnd: external,
+				protocol: proto,
+				host,
+			});
+		});
+
+		return PortMap.normalisePortMaps(portMaps);
+	}
+
+	public static normalisePortMaps(portMaps: PortMap[]): PortMap[] {
+		// Fold any ranges into each other if possible
+		return _(portMaps)
+			.sortBy((p) => p.protocol)
+			.sortBy((p) => p.host)
+			.sortBy((p) => p.internalStart)
+			.reduce((res: PortMap[], p: PortMap) => {
+				const last = _.last(res);
+				if (last == null) {
```

Comment:
You could remove this by using
```
if (last != null &&
	last.internalEnd + 1 === p.internalStart &&
	...
) {
...
} else {
	res.push(p)
}
```
so you just have the one path for `res.push(p)`